File: plugins/wise/engine/src/dispatch.ts
// `models` and `dispatch` CLI commands: the catalog for skills that offer a harness/model/effort
// pick, and a one-off child run on any harness — how a skill (e.g. /wise-pr-watch-auto --on codex)
// runs its own procedure as a subagent of another CLI without going through a workflow.

import { readFileSync } from "node:fs";
import { adapterFor, hasAdapter } from "./adapters/index.ts";
import { catalogFor, catalogModel, defaultModel } from "./models.ts";
import { headline } from "./steps/agent.ts";
import { EFFORTS, HARNESSES, RUN_MODES } from "./types.ts";
import type { Adapter, Effort, Harness, RunMode, RunReq, RunRes } from "./types.ts";

export type DispatchIo = { out: (s: string) => void; err: (s: string) => void };

type Flags = Record<string, string | true>;

function str(flag: string | true | undefined): string | undefined {
  return typeof flag === "string" ? flag : undefined;
}

function isHarness(s: string): s is Harness {
  return (HARNESSES as readonly string[]).includes(s);
}

/**
 * `models [harness...] [--text]`: the model catalog per harness — id, label, description and the
 * efforts each model takes. Skills read this instead of hardcoding a list.
 */
export function cmdModels(positional: string[], flags: Flags, io: DispatchIo): number {
  const wanted = positional.length > 0 ? positional : [...HARNESSES];
  const rows: {
    harness: Harness;
    id: string;
    label: string;
    description: string;
    efforts: readonly Effort[];
  }[] = [];
  for (const name of wanted) {
    if (!isHarness(name)) {
      io.err(`models: unknown harness ${name} (one of ${HARNESSES.join(", ")})\n`);
      return 2;
    }
    for (const m of catalogFor(name)) rows.push({ harness: name, ...m });
  }
  if (flags.text === true) {
    for (const r of rows) {
      io.out(
        `${r.harness}\t${r.id}\t${r.label}\t${r.efforts.length > 0 ? r.efforts.join(",") : "-"}\t${r.description}\n`,
      );
    }
  } else {
    io.out(JSON.stringify(rows) + "\n");
  }
  return 0;
}

export type DispatchResult = {
  ok: boolean;
  exit: RunRes["exit"];
  harness: Harness;
  model: string;
  effort: Effort | null;
  mode: RunMode;
  verdict: string;
  text: string;
  usage: RunRes["usage"];
  error?: string;
  warnings: string[];
};

const DEFAULT_DISPATCH_TIMEOUT_S = 3600;

/**
 * `dispatch --harness <h> --prompt-file <path> [--model <id>] [--effort <e>] [--mode <m>]
 * [--cwd <dir>] [--timeout-s <n>] [--add-dir <dir>] [--allowed-tools <a,b>] [--text]`
 *
 * One child run, no daemon, no ledger: read the prompt, start the harness, wait, print one JSON
 * object (or the child's text under --text). Exit 0 on a clean child, 1 otherwise. A model off
 * the catalog passes through as typed; an effort the model does not list is an error, so a typo
 * never silently changes the spend.
 */
export async function cmdDispatch(
  flags: Flags,
  io: DispatchIo,
  adapters: (h: Harness) => Adapter = adapterFor,
): Promise<number> {
  const harnessRaw = str(flags.harness);
  if (harnessRaw === undefined || !isHarness(harnessRaw)) {
    io.err(`dispatch: --harness must be one of ${HARNESSES.join(", ")}\n`);
    return 64;
  }
  const harness = harnessRaw;
  if (!hasAdapter(harness)) {
    io.err(`dispatch: no adapter for harness ${harness}\n`);
    return 64;
  }
  const promptFile = str(flags["prompt-file"]);
  const promptInline = str(flags.prompt);
  if (promptFile === undefined && promptInline === undefined) {
    io.err("dispatch: --prompt-file <path> (or --prompt <text>) is required\n");
    return 64;
  }
  let prompt: string;
  try {
    prompt = promptInline ?? readFileSync(promptFile as string, "utf8");
  } catch (e) {
    io.err(`dispatch: cannot read prompt file: ${(e as Error).message}\n`);
    return 66;
  }
  if (prompt.trim().length === 0) {
    io.err("dispatch: prompt is empty\n");
    return 64;
  }

  const warnings: string[] = [];
  const modelFlag = str(flags.model);
  const inCatalog = catalogModel(harness, modelFlag);
  const model = modelFlag !== undefined ? (inCatalog?.id ?? modelFlag) : defaultModel(harness).id;
  if (modelFlag !== undefined && inCatalog === undefined) {
    warnings.push(`model ${modelFlag} is not in the catalog; passed through as typed`);
  }

  const effortFlag = str(flags.effort);
  let effort: Effort | undefined;
  if (effortFlag !== undefined) {
    if (!(EFFORTS as readonly string[]).includes(effortFlag)) {
      io.err(`dispatch: --effort must be one of ${EFFORTS.join(", ")}\n`);
      return 64;
    }
    const known = inCatalog ?? (modelFlag === undefined ? defaultModel(harness) : undefined);
    if (known && !known.efforts.includes(effortFlag as Effort)) {
      io.err(
        `dispatch: model ${known.id} takes ${known.efforts.length > 0 ? known.efforts.join(", ") : "no effort flag"}, not ${effortFlag}\n`,
      );
      return 64;
    }
    effort = effortFlag as Effort;
  }

  const modeFlag = str(flags.mode) ?? "auto";
  if (!(RUN_MODES as readonly string[]).includes(modeFlag)) {
    io.err(`dispatch: --mode must be one of ${RUN_MODES.join(", ")}\n`);
    return 64;
  }
  const mode = modeFlag as RunMode;

  const timeoutS = str(flags["timeout-s"]);
  const timeoutMs = (timeoutS !== undefined ? Number(timeoutS) : DEFAULT_DISPATCH_TIMEOUT_S) * 1000;
  if (!Number.isFinite(timeoutMs) || timeoutMs <= 0) {
    io.err("dispatch: --timeout-s must be a positive number\n");
    return 64;
  }

  const req: RunReq = {
    prompt,
    model,
    cwd: str(flags.cwd) ?? process.cwd(),
    mode,
    timeout_ms: timeoutMs,
    auth: "subscription",
    ...(effort !== undefined ? { effort } : {}),
    ...(str(flags["add-dir"]) !== undefined ? { add_dirs: [str(flags["add-dir"]) as string] } : {}),
    ...(str(flags["allowed-tools"]) !== undefined
      ? { allowed_tools: (str(flags["allowed-tools"]) as string).split(",").filter(Boolean) }
      : {}),
  };

  const res = await adapters(harness).run(req, () => {});
  const result: DispatchResult = {
    ok: res.exit === "ok",
    exit: res.exit,
    harness,
    model: res.model ?? model,
    effort: effort ?? null,
    mode,
    verdict: headline(res.text),
    text: res.text,
    usage: res.usage,
    ...(res.error !== undefined ? { error: res.error } : {}),
    warnings: [...warnings, ...(res.warnings ?? [])],
  };
  if (flags.text === true) {
    io.out(res.text.endsWith("\n") || res.text.length === 0 ? res.text : res.text + "\n");
    if (!result.ok)
      io.err(`dispatch: child exit ${res.exit}${res.error ? `: ${res.error}` : ""}\n`);
  } else {
    io.out(JSON.stringify(result) + "\n");
  }
  return result.ok ? 0 : 1;
}

File: plugins/wise/engine/wise_engine/dispatch.py
from __future__ import annotations

import math
import os
from collections.abc import Callable, Mapping
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from .adapter_types import Adapter, Json
from .adapters import adapter_for, has_adapter
from .adapters._common import dumps
from .constants import EFFORTS, HARNESSES, RUN_MODES
from .models import catalog_model, default_model, discover_models, merged_catalog
from .steps.agent import headline, project_system_prompt


@dataclass
class DispatchIo:
    out: Callable[[str], Any]
    err: Callable[[str], Any]


def _string(value: Any) -> str | None:
    return value if isinstance(value, str) else None


async def cmd_models(
    positional: list[str],
    flags: Json,
    io: DispatchIo,
    env: Mapping[str, str] | None = None,
    lookup: Callable[[str], Any] | None = None,
) -> int:
    from .auth import bin_on_path

    names = list(positional or HARNESSES)
    for name in names:
        if name not in HARNESSES:
            io.err(f"models: unknown harness {name} (one of {', '.join(HARNESSES)})\n")
            return 2
    discovered: Json = {}
    if flags.get("catalog-only") is not True:
        lookup = lookup or (lambda name: adapter_for(name) if has_adapter(name) else None)
        installed = []
        for name in names:
            adapter = lookup(name)
            if adapter is None:
                continue
            binary = getattr(adapter, "bin", None)
            if not isinstance(binary, str) or bin_on_path(binary, env):
                installed.append(name)
        discovered = await discover_models(installed, lookup)
    rows = []
    for name in names:
        rows += [
            dict(harness=name, **model) for model in merged_catalog(name, discovered.get(name))
        ]
    if flags.get("text") is True:
        for row in rows:
            io.out(
                f"{row['harness']}\t{row['id']}\t{row['label']}\t{','.join(row['efforts']) or '-'}\t{row['source']}\t{row['description']}\n"
            )
    else:
        io.out(dumps(rows) + "\n")
    return 0


def prepare_dispatch(flags: Json, io: DispatchIo) -> tuple[str, Json, list[str]] | int:
    harness = _string(flags.get("harness"))
    if harness is None or harness not in HARNESSES:
        io.err(f"dispatch: --harness must be one of {', '.join(HARNESSES)}\n")
        return 64
    if not has_adapter(harness):
        io.err(f"dispatch: no adapter for harness {harness}\n")
        return 64
    prompt_file, prompt_inline = _string(flags.get("prompt-file")), _string(flags.get("prompt"))
    if prompt_file is None and prompt_inline is None:
        io.err("dispatch: --prompt-file <path> (or --prompt <text>) is required\n")
        return 64
    try:
        prompt = (
            prompt_inline
            if prompt_inline is not None
            else Path(prompt_file or "").read_text(encoding="utf-8")
        )
    except OSError as error:
        io.err(f"dispatch: cannot read prompt file: {error}\n")
        return 66
    if not prompt.strip():
        io.err("dispatch: prompt is empty\n")
        return 64
    warnings = []
    model_flag = _string(flags.get("model"))
    in_catalog = catalog_model(harness, model_flag)
    model = (
        (in_catalog["id"] if in_catalog else model_flag)
        if model_flag is not None
        else default_model(harness)["id"]
    )
    if model_flag is not None and in_catalog is None:
        warnings.append(f"model {model_flag} is not in the catalog; passed through as typed")
    effort = _string(flags.get("effort"))
    if effort is not None:
        if effort not in EFFORTS:
            io.err(f"dispatch: --effort must be one of {', '.join(EFFORTS)}\n")
            return 64
        known = in_catalog or (default_model(harness) if model_flag is None else None)
        if known and effort not in known["efforts"]:
            io.err(
                f"dispatch: model {known['id']} takes {', '.join(known['efforts']) or 'no effort flag'}, not {effort}\n"
            )
            return 64
    mode = _string(flags.get("mode"))
    if mode is None:
        mode = "auto"
    if mode not in RUN_MODES:
        io.err(f"dispatch: --mode must be one of {', '.join(RUN_MODES)}\n")
        return 64
    timeout_s = _string(flags.get("timeout-s"))
    try:
        timeout_ms = (float(timeout_s.strip() or "0") if timeout_s is not None else 3600) * 1000
    except ValueError:
        timeout_ms = math.nan
    if not math.isfinite(timeout_ms) or timeout_ms <= 0:
        io.err("dispatch: --timeout-s must be a positive number\n")
        return 64
    cwd_flag = flags.get("cwd")
    cwd = cwd_flag if isinstance(cwd_flag, str) else os.getcwd()
    req = dict(
        prompt=prompt,
        model=model,
        cwd=cwd,
        mode=mode,
        timeout_ms=timeout_ms,
        auth="subscription",
    )
    req["system"] = project_system_prompt(cwd)
    if effort is not None:
        req["effort"] = effort
    if isinstance(flags.get("add-dir"), str):
        req["add_dirs"] = [flags["add-dir"]]
    if isinstance(flags.get("allowed-tools"), str):
        req["allowed_tools"] = [tool for tool in flags["allowed-tools"].split(",") if tool]
    return harness, req, warnings


async def cmd_dispatch(
    flags: Json, io: DispatchIo, adapters: Callable[[str], Adapter] = adapter_for
) -> int:
    prepared = prepare_dispatch(flags, io)
    if isinstance(prepared, int):
        return prepared
    harness, req, warnings = prepared
    model, effort, mode = req["model"], req.get("effort"), req["mode"]
    res = await adapters(harness).run(req, lambda event: None)
    result = dict(
        ok=res["exit"] == "ok",
        exit=res["exit"],
        harness=harness,
        model=res.get("model", model),
        effort=effort,
        mode=mode,
        verdict=headline(res["text"]),
        text=res["text"],
        usage=res["usage"],
        warnings=[*warnings, *res.get("warnings", [])],
    )
    if "error" in res:
        result["error"] = res["error"]
    if flags.get("text") is True:
        io.out(res["text"] if not res["text"] or res["text"].endswith("\n") else res["text"] + "\n")
        if not result["ok"]:
            io.err(
                f"dispatch: child exit {res['exit']}{': ' + res['error'] if res.get('error') else ''}\n"
            )
    else:
        io.out(dumps(result) + "\n")
    return 0 if result["ok"] else 1
